src/Microsoft.OData.Core/ODataBatchWriter.cs

```diff
@@ -422,314 +392,202 @@ void IODataBatchOperationListener.BatchOperationContentStreamDisposed()
-        }
+        /// <returns>Task representing the pending flush operation.</returns>
+        protected abstract Task FlushAsynchronously();
+#endif
 
         /// <summary>
-        /// Ends a batch - implementation of the actual functionality.
+        /// Ends a batch.
         /// </summary>
-        private void WriteEndBatchImplementation()
-        {
-            Debug.Assert(
-                this.batchStartBoundaryWritten || this.CurrentOperationMessage == null,
-                "If not batch boundary was written we must not have an active message.");
+        protected abstract void WriteEndBatchImplementation();
 
-            // write pending message data (headers, response line) for a previously unclosed message/request
-            this.WritePendingMessageData(true);
+        /// <summary>
+        /// Starts a new changeset.
+        /// </summary>
+        protected abstract void WriteStartChangesetImplementation();
 
-            this.SetState(BatchWriterState.BatchCompleted);
+        /// <summary>
+        /// Ends an active changeset.
+        /// </summary>
+        protected abstract void WriteEndChangesetImplementation();
 
-            // write the end boundary for the batch
-            ODataBatchWriterUtils.WriteEndBoundary(this.rawOutputContext.TextWriter, this.batchBoundary, !this.batchStartBoundaryWritten);
+        /// <summary>
+        /// Creates an <see cref="ODataBatchOperationResponseMessage"/> for writing an operation of a batch response.
+        /// </summary>
+        /// <param name="contentId">The Content-ID value to write in ChangeSet head.</param>
+        /// <returns>The message that can be used to write the response operation.</returns>
+        protected abstract ODataBatchOperationResponseMessage CreateOperationResponseMessageImplementation(string contentId);
 
-            // For compatibility with WCF DS we write a newline after the end batch boundary.
-            // Technically it's not needed, but it doesn't violate anything either.
-            this.rawOutputContext.TextWriter.WriteLine();
-        }
+        /// <summary>
+        /// Writes all the pending headers and prepares the writer to write a content of the operation.
+        /// </summary>
+        protected abstract void StartBatchOperationContent();
 
         /// <summary>
-        /// Verifies that calling WriteStartChangeset is valid.
+        /// Whether the writer is currently processing inside a sub-batch (changeset or atomic group).
         /// </summary>
-        /// <param name="synchronousCall">true if the call is to be synchronous; false otherwise.</param>
-        private void VerifyCanWriteStartChangeset(bool synchronousCall)
-        {
-            this.ValidateWriterReady();
-            this.VerifyCallAllowed(synchronousCall);
-        }
+        /// <returns>True if the writer processing is inside a sub-batch.</returns>
+        protected abstract bool IsInsideSubBatch();

```

Comment:
We should remove this protected method; the base class should not need to call into derived class to see if it is in a changeset; it should keep track of this through state transitions (i.e., if WriteStartChangeset has been called without an accompanying WriteEndChangeset.)

Reply:
Updated to use private method in base class: return this.state == BatchWriterState.ChangesetStarted;

---
In reply to: [143605390](https://github.com/OData/odata.net/pull/960#discussion_r143605390) [](ancestors = 143605390)